pkg/eventrecorder/eventrecorder.go

```diff
@@ -86,61 +82,40 @@ type EventDetailsError struct {
-	evt.EventDetails = &EventDetailsError{errorString}
-	er.recordEvent("QueryFailure", evt)
-}
-
-// QuerySuccess events occur on successfully querying a storage
-// provider. A query will result in either a QueryFailure or
-// a QuerySuccess event.
-func (er *EventRecorder) QuerySuccess(retrievalId types.RetrievalID, phaseStartTime, eventTime time.Time, retrievalCid cid.Cid, storageProviderId peer.ID, queryResponse retrievalmarket.QueryResponse) {
-	evt := eventReport{retrievalId, er.instanceId, retrievalCid.String(), storageProviderId, eventpublisher.QueryPhase, phaseStartTime, eventpublisher.QueryAskedCode, eventTime, nil}
-	evt.EventDetails = &queryResponse
-	er.recordEvent("QuerySuccess", evt)
-}
+func (er *EventRecorder) RecordEvent(event types.RetrievalEvent) {
+	if event.Phase() == types.IndexerPhase {
+		// ignore indexer events for now, it can get very chatty in the autoretrieve
+		// case where every request results in an indexer lookup
+		return
+	}
 
-// RetrievalProgress events occur during the process of a retrieval. The
-// Success and failure progress event types are not reported here, but are
-// signalled via RetrievalSuccess or RetrievalFailure.
-func (er *EventRecorder) RetrievalProgress(retrievalId types.RetrievalID, phaseStartTime, eventTime time.Time, retrievalCid cid.Cid, storageProviderId peer.ID, eventName eventpublisher.Code) {
-	evt := eventReport{retrievalId, er.instanceId, retrievalCid.String(), storageProviderId, eventpublisher.RetrievalPhase, phaseStartTime, eventName, eventTime, nil}
-	er.recordEvent("RetrievalProgress", evt)
-}
+	evt := eventReport{
+		RetrievalId:       event.RetrievalId(),
+		InstanceId:        er.instanceId,
+		Cid:               event.PayloadCid().String(),
+		StorageProviderId: event.StorageProviderId(),
+		Phase:             event.Phase(),
+		PhaseStartTime:    event.PhaseStartTime(),
+		EventName:         event.Code(),
+		EventTime:         event.Time(),
+	}
 
-// RetrievalSuccess events occur on the success of a retrieval. A retrieval
-// will result in either a QueryFailure or a QuerySuccess
-// event.
-func (er *EventRecorder) RetrievalSuccess(retrievalId types.RetrievalID, phaseStartTime, eventTime time.Time, retrievalCid cid.Cid, storageProviderId peer.ID, retrievedSize uint64, receivedCids uint64, confirmed bool) {
-	evt := eventReport{retrievalId, er.instanceId, retrievalCid.String(), storageProviderId, eventpublisher.RetrievalPhase, phaseStartTime, eventpublisher.SuccessCode, eventTime, nil}
-	evt.EventDetails = &EventDetailsSuccess{retrievedSize, receivedCids, confirmed}
-	er.recordEvent("RetrievalSuccess", evt)
-}
+	switch ret := event.(type) {
+	case events.EventWithCandidates: // events.RetrievalEventCandidatesFound, events.RetrievalEventCandidatesFiltered:
+	case events.RetrievalEventConnected:
+	case events.EventWithQueryResponse: // events.RetrievalEventQueryAsked, events.RetrievalEventQueryAskedFiltered:
```

Comment:
huh! I didn't know you could switch an interface to an interface but I guess you can.

Reply:
yeah, kind of weird, but works: https://go.dev/play/p/2LbRnZs0BU
this branch also shows up in coverage, so it's being exercised